src/NSwag.AspNetCore/NSwagDocumentProvider.cs

```diff
@@ -0,0 +1,116 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="NSwagDocumentProvider.cs" company="NSwag">
+//     Copyright (c) Rico Suter. All rights reserved.
+// </copyright>
+// <license>https://github.com/NSwag/NSwag/blob/master/LICENSE.md</license>
+// <author>Rico Suter, mail@rsuter.com</author>
+//-----------------------------------------------------------------------
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.ApiExplorer;
+using Microsoft.Extensions.ApiDescription;
+using Microsoft.Extensions.Options;
+using NSwag.SwaggerGeneration;
+using NSwag.SwaggerGeneration.AspNetCore;
+using NSwag.SwaggerGeneration.WebApi;
+
+namespace NSwag.AspNetCore
+{
+    internal class NSwagDocumentProvider : IDocumentProvider
+    {
+        private readonly IApiDescriptionGroupCollectionProvider _apiDescriptionGroupCollectionProvider;
+        private readonly IOptions<MvcOptions> _mvcOptions;
+        private readonly IOptions<MvcJsonOptions> _mvcJsonOptions;
+        private readonly DocumentRegistry _registry;
+
+        public NSwagDocumentProvider(
+            DocumentRegistry registry,
+            IApiDescriptionGroupCollectionProvider apiDescriptionGroupCollectionProvider,
+            IOptions<MvcOptions> mvcOptions,
+            IOptions<MvcJsonOptions> mvcJsonOptions)
+        {
+            if (apiDescriptionGroupCollectionProvider == null)
+            {
+                throw new ArgumentNullException(nameof(apiDescriptionGroupCollectionProvider));
+            }
+
+            if (mvcOptions == null)
+            {
+                throw new ArgumentNullException(nameof(mvcOptions));
+            }
+
+            if (mvcJsonOptions == null)
+            {
+                throw new ArgumentNullException(nameof(mvcJsonOptions));
+            }
+
+            if (registry == null)
+            {
+                throw new ArgumentNullException(nameof(registry));
+            }
+
+            _apiDescriptionGroupCollectionProvider = apiDescriptionGroupCollectionProvider;
+            _mvcOptions = mvcOptions;
+            _mvcJsonOptions = mvcJsonOptions;
+            _registry = registry;
+        }
+
+        public async Task<SwaggerDocument> GenerateAsync(string documentName)
+        {
+            if (documentName == null)
```

Comment:
Maybe it would be better to allow null and throw an exception when more than one document is added with null? This way a single document - probably the most used scenario - is automatically added.

Reply:
Te Microsoft tool won't pass `null` into `IDocumentProvider.GenerateAsync(...)` and instead uses `"v1"` by default. However, I could change registration to use `"v1"` when the new `DocumentName` property is `null`. Thoughts?

Separately, the registrar currently has a "last one wins" semantic. Not sure we should require distinct `DocumentName` properties in the initial release. Wouldn't that be a breaking change?

Reply:
That would be a breaking change.

Reply:
Probably this should be a breaking change in v12. Good thing is that it's easy to communicate how to fix (exceptions). #984

Reply:
To clarify the resolution for this PR, I believe I should
1. Leave this `null` check alone
2. Change the registrations to use `"v1"` when `DocumentName` isn't set
3. Leave the "last one wins" semantic alone pending work in the v12 epic

On (2), would you prefer the `DocumentName` property simply have a default value of `"v1"`?

Reply:
I think we should do 2 and 3 and just use a default value of "v1"

Reply:
Done